importers/rules.go

```diff
@@ -0,0 +1,256 @@
+		}
+
+		for _, ruleGroup := range ruleGroups.Groups {
+			itv := importer.config.EvalInterval
+			if ruleGroup.Interval != 0 {
+				itv = time.Duration(ruleGroup.Interval)
+			}
+
+			rulez := make([]rules.Rule, 0, len(ruleGroup.Rules))
+			for _, r := range ruleGroup.Rules {
+				expr, err := parser.ParseExpr(r.Expr.Value)
+				if err != nil {
+					return []error{errors.Wrap(err, file)}
+				}
+
+				rulez = append(rulez, rules.NewRecordingRule(
+					r.Record.Value,
+					expr,
+					labels.FromMap(r.Labels),
+				))
+			}
+
+			groups[file+";"+ruleGroup.Name] = rules.NewGroup(rules.GroupOptions{
+				Name:     ruleGroup.Name,
+				File:     file,
+				Interval: itv,
+				Rules:    rulez,
+			})
+		}
+	}
+
+	importer.groups = groups
+	return errs
+}
+
+// ImportAll evaluates all the groups and rules and creates new time series
+// and stores in new blocks
+func (importer *RuleImporter) ImportAll(ctx context.Context) []error {
+	var errs = []error{}
+	for _, group := range importer.groups {
+		for _, rule := range group.Rules() {
+			err := importer.ImportRule(ctx, rule)
+			if err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+	err := importer.CreateBlocks()
+	if err != nil {
+		errs = append(errs, err)
+	}
+	return errs
+}
+
+func (importer *RuleImporter) queryFn(ctx context.Context, q string, t time.Time) (promql.Vector, error) {
+	val, warnings, err := importer.apiClient.Query(ctx, q, t)
+	if err != nil {
+		return promql.Vector{}, err
+	}
+	if warnings != nil {
+		fmt.Fprint(os.Stderr, "warning api.Query:", warnings)
+	}
+
+	switch val.Type() {
+	case model.ValVector:
+		valVector := val.(model.Vector)
+		return modelToPromqlVector(valVector), nil
+	case model.ValScalar:
+		valScalar := val.(*model.Scalar)
+		return promql.Vector{promql.Sample{
+			Metric: labels.Labels{},
+			Point:  promql.Point{T: int64(valScalar.Timestamp), V: float64(valScalar.Value)},
+		}}, nil
+	default:
+		return nil, errors.New("rule result is wrong type")
+	}
+}
+
+func modelToPromqlVector(modelValue model.Vector) promql.Vector {
+	result := make(promql.Vector, 0, len(modelValue))
+
+	for _, value := range modelValue {
+		labels := make(labels.Labels, 0, len(value.Metric))
+
+		for k, v := range value.Metric {
+			labels = append(labels, plabels.Label{
+				Name:  string(k),
+				Value: string(v),
+			})
+		}
+		sort.Sort(labels)
+
+		result = append(result, promql.Sample{
+			Metric: labels,
+			Point:  promql.Point{T: int64(value.Timestamp), V: float64(value.Value)},
+		})
+	}
+	return result
+}
+
+// ImportRule imports the historical data for a single rule
+func (importer *RuleImporter) ImportRule(ctx context.Context, rule rules.Rule) error {
+	ts, err := parseTime(importer.config.Start)
+	if err != nil {
+		return err
+	}
+	end, err := parseTime(importer.config.End)
+	if err != nil {
+		return err
+	}
+	url, err := url.Parse(importer.config.URL)
+	if err != nil {
+		return err
+	}
+
+	appender := importer.writer.Appender()
+	for ts.Before(end) {
```

Comment:
You're still using the user-specified time for alignment, which won't work out.

Reply:
Yes, we just talked about this.

The idea would be starting from zero if no data is yet in Prom
But actually we could use instead of 0, a group hash as in Prometheus. The tricky part is a filename, as it obviously will be different.

```
// evalTimestamp returns the immediately preceding consistently slotted evaluation time.
func (g *Group) evalTimestamp() time.Time {
	var (
		offset = int64(g.hash() % uint64(g.interval))
		now    = time.Now().UnixNano()
		adjNow = now - offset
		base   = adjNow - (adjNow % int64(g.interval))
	)

	return time.Unix(0, base+offset).UTC()
}
```
If the data exists we can fetch samples from Prometheus and aligned (:

Reply:
I exported the evalTimestamp on the Group and used that in my code for the alignment. Let me know if you see any glaring problems with that. I still need to test it, which I'm going to do today.